tools/clang/test/SemaHLSL/recursion/library_recursion_unexported.hlsl

```diff
@@ -0,0 +1,59 @@
+// RUN: %dxc -Tlib_6_5 -verify %s 
+// This file tests several things at once:
+// Firstly, recursive functions with external linkage should be tested
+// and diagnosed for recursion. So, exported_recurse should throw an error
+// since it is recursive. 
+// Secondly, functions with external linkage may not call recursive 
+// functions with internal linkage. exported_recurse_2 calls 
+// unexported_recurse, so we expect an error.
+// Thirdly, unreachable functions with internal linkage should not 
+// be tested for recursion in library shaders. 
+// In this case, unreachable_unexported_recurse is an example, and we 
+// do not expect any compilation errors for that function.
+// Finally, functions with unspecified linkage will default to external
+// for library shaders, so we expect errors on unreachable functions
+// that are recursive
+
+// expected-error@+1{{recursive functions are not allowed: export function calls recursive function 'unreachable_unexported_recurse_external'}}
```

Comment:
There is no export function which called unreachable_unexported_recurse_external ?

Reply:
My understanding is that this function has external linkage, so it will be exported. The function calls itself, so the diagnostic is accurate. Is this incorrect?

Reply:
Only export function be exported unless -default-linkage is set to external.
We need add hlsl spec about library profile to make things clear.
The name 'external' is really make things hard to understand here :(